File: ios/WalletKit/Sources/WalletKit/WalletClient.swift
import Foundation

// The gomobile-generated bindings. The framework name comes from the
// `make mobile-ios` output (Wavewalletdk.xcframework), and the symbol prefix is the
// Go package name, "Mobile" (e.g. MobileStart, MobileGetInfo, MobileSubscribe).
// All generated-symbol use is confined to the `Bindings` enum below, so if the
// gomobile prefix changes it is a one-place edit.
#if canImport(Wavewalletdk)
import Wavewalletdk
#endif

/// Thrown when an embedded-wallet call fails. Wraps the Go-side error.
public struct WalletError: Error, Sendable {
    public let message: String
    init(_ error: Error) { self.message = (error as NSError).localizedDescription }
    init(message: String) { self.message = message }
}

/// An idiomatic Swift facade over the gomobile bindings.
///
/// Every call is `async throws`, runs the blocking binding off the main actor,
/// and returns a typed model. The wallet activity stream is an
/// `AsyncThrowingStream` that closes its subscription when the consuming task
/// ends. One embedded daemon runs per process, so a single shared instance is
/// enough for the whole app.
public actor WalletClient {
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    public init() {
        decoder = JSONDecoder()
        encoder = JSONEncoder()
    }

    /// Whether the embedded daemon is currently running.
    public var isRunning: Bool { Bindings.isRunning() }

    /// Boot the embedded daemon and wait until its gRPC channel is serving.
    public func start(_ config: WalletConfig) async throws {
        let json = String(decoding: try encoder.encode(config), as: UTF8.self)
        try await bg { try Bindings.start(json) }
    }

    /// Tear the daemon down. Idempotent.
    public func stop() async throws {
        try await bg { try Bindings.stop() }
    }

    /// Daemon readiness snapshot.
    public func getInfo() async throws -> Info {
        try decode(await bg { try Bindings.getInfo() })
    }

    /// Wallet readiness, balance, and pending-activity count.
    public func status() async throws -> Status {
        try decode(await bg { try Bindings.status() })
    }

    /// Wallet balance summary.
    public func balance() async throws -> Balance {
        try decode(await bg { try Bindings.balance() })
    }

    /// Create or import the wallet. An empty mnemonic generates a fresh seed.
    public func createWallet(
        walletPassword: Data,
        mnemonic: [String] = []
    ) async throws -> CreateWalletResult {
        let req = CreateWalletReq(
            mnemonic: mnemonic,
            walletPassword: walletPassword.base64EncodedString()
        )
        let body = try encoder.encode(req)
        return try decode(await bg { try Bindings.createWallet(body) })
    }

    /// Unlock an existing wallet.
    public func unlockWallet(walletPassword: Data) async throws -> UnlockWalletResult {
        let req = UnlockWalletReq(walletPassword: walletPassword.base64EncodedString())
        let body = try encoder.encode(req)
        return try decode(await bg { try Bindings.unlockWallet(body) })
    }

    /// Open a Lightning invoice to receive `amountSat` into the wallet. Share
    /// the returned invoice with the payer; the matching entry shows up on the
    /// activity stream as it is paid.
    public func receiveLightning(amountSat: Int64, memo: String = "") async throws -> ReceiveResult {
        let body = try encoder.encode(ReceiveReq(amountSat: amountSat, memo: memo))
        return try decode(await bg { try Bindings.receive(body) })
    }

    /// Allocate a fresh on-chain boarding address to deposit into.
    public func newDepositAddress(amountSatHint: Int64 = 0) async throws -> DepositResult {
        let body = try encoder.encode(DepositReq(amountSatHint: amountSatHint))
        return try decode(await bg { try Bindings.deposit(body) })
    }

    /// Quote an outbound payment without moving funds. Set exactly one of
    /// `invoice` or `onchainAddress`. The result carries a single-use intent id
    /// to pass to `send(_:)`. `sweepAll` drains every VTXO to the address.
    public func prepareSend(
        invoice: String? = nil,
        onchainAddress: String? = nil,
        amountSat: Int64 = 0,
        note: String = "",
        maxFeeSat: Int64 = 0,
        sweepAll: Bool = false
    ) async throws -> PrepareSendResult {
        let req = PrepareSendReq(
            invoice: invoice, onchainAddress: onchainAddress, amountSat: amountSat,
            note: note, maxFeeSat: maxFeeSat, sweepAll: sweepAll
        )
        let body = try encoder.encode(req)
        return try decode(await bg { try Bindings.prepareSend(body) })
    }

    /// Dispatch a prepared send, consuming the single-use intent id.
    public func send(_ sendIntentID: String) async throws -> SendResult {
        let body = try encoder.encode(SendPreparedReq(sendIntentID: sendIntentID))
        return try decode(await bg { try Bindings.sendPrepared(body) })
    }

    /// Pay a Lightning invoice in one call, skipping the explicit quote step.
    /// Use `prepareSend` + `send` when you want to show the fee before paying.
    public func payInvoice(_ invoice: String, maxFeeSat: Int64 = 0) async throws -> SendResult {
        let quote = try await prepareSend(invoice: invoice, maxFeeSat: maxFeeSat)
        return try await send(quote.sendIntentID)
    }

    /// List a unified wallet view: the merged activity history, the live VTXO
    /// inventory, or the on-chain transaction history. Read the `ListResult`
    /// property named by its `view`. `kinds`, `pendingOnly`, `limit`, and
    /// `offset` apply to the activity view.
    public func list(
        view: ListView = .activity,
        kinds: [String] = [],
        pendingOnly: Bool = false,
        limit: Int64 = 0,
        offset: Int64 = 0
    ) async throws -> ListResult {
        let req = ListReq(
            view: view.rawValue, pendingOnly: pendingOnly,
            kinds: kinds, limit: limit, offset: offset
        )
        let body = try encoder.encode(req)
        return try decode(await bg { try Bindings.list(body) })
    }

    /// Exit a VTXO back to the chain. The daemon queues a cooperative leave by
    /// default, paying out to `destination` (or a fresh backing-wallet address
    /// when empty). To bypass cooperation and start a unilateral unroll, pass
    /// `forceUnrollAck` = "I_KNOW_WHAT_I_AM_DOING"; it cannot be combined with
    /// `destination`. Track progress with `exitStatus` or the activity stream.
    public func exit(
        outpoint: String,
        destination: String = "",
        forceUnrollAck: String = ""
    ) async throws -> ExitResult {
        let req = ExitReq(
            outpoint: outpoint, destination: destination,
            forceUnrollAck: forceUnrollAck
        )
        let body = try encoder.encode(req)
        return try decode(await bg { try Bindings.exit(body) })
    }

    /// Query the phase of an exit job for a VTXO outpoint.
    public func exitStatus(outpoint: String) async throws -> ExitStatusResult {
        let body = try encoder.encode(ExitStatusReq(outpoint: outpoint))
        return try decode(await bg { try Bindings.exitStatus(body) })
    }

    /// Stream only incoming payments (receives and deposits) as they arrive and
    /// settle. A convenience over `activity` filtered to the credit kinds: watch
    /// for an `Entry` whose status becomes "complete".
    public nonisolated func incomingPayments(
        includeExisting: Bool = false
    ) -> AsyncThrowingStream<Entry, Error> {
        activity(includeExisting: includeExisting, kinds: ["receive", "deposit"])
    }

    /// Stream wallet activity. The stream finishes on a clean end-of-stream and
    /// throws `WalletError` on a real error. Cancelling the consuming task
    /// closes the underlying subscription, which unblocks the pull loop.
    public nonisolated func activity(
        includeExisting: Bool = false,
        kinds: [String] = []
    ) -> AsyncThrowingStream<Entry, Error> {
        AsyncThrowingStream { continuation in
            let decoder = JSONDecoder()
            let reqData = (try? JSONEncoder().encode(
                SubscribeReq(includeExisting: includeExisting, kinds: kinds)
            )) ?? Data("{}".utf8)

            let sub: BindingsSubscription
            do {
                sub = try Bindings.subscribe(reqData)
            } catch {
                continuation.finish(throwing: WalletError(error))
                return
            }

            let task = Task.detached {
                while !Task.isCancelled {
                    let data: Data
                    do {
                        data = try sub.next()
                    } catch {
                        if Self.isEndOfStream(error) { break }
                        continuation.finish(throwing: WalletError(error))
                        return
                    }
                    if let entry = try? decoder.decode(Entry.self, from: data) {
                        continuation.yield(entry)
                    }
                }
                continuation.finish()
            }

            continuation.onTermination = { _ in
                try? sub.close()
                task.cancel()
            }
        }
    }

    // MARK: - internals

    private func bg<T: Sendable>(_ work: @Sendable @escaping () throws -> T) async throws -> T {
        try await withCheckedThrowingContinuation { cont in
            DispatchQueue.global(qos: .userInitiated).async {
                do { cont.resume(returning: try work()) }
                catch { cont.resume(throwing: WalletError(error)) }
            }
        }
    }

    private func decode<T: Decodable>(_ data: Data) throws -> T {
        do { return try decoder.decode(T.self, from: data) }
        catch { throw WalletError(error) }
    }

    private static func isEndOfStream(_ error: Error) -> Bool {
        (error as NSError).localizedDescription.lowercased().contains("eof")
    }
}

// Request bodies. Keys match the Go DTO field names (PascalCase).
private struct CreateWalletReq: Encodable {
    let mnemonic: [String]
    let walletPassword: String
    enum CodingKeys: String, CodingKey {
        case mnemonic = "Mnemonic"
        case walletPassword = "WalletPassword"
    }
}

private struct UnlockWalletReq: Encodable {
    let walletPassword: String
    enum CodingKeys: String, CodingKey {
        case walletPassword = "WalletPassword"
    }
}

private struct SubscribeReq: Encodable {
    let includeExisting: Bool
    let kinds: [String]
    enum CodingKeys: String, CodingKey {
        case includeExisting = "IncludeExisting"
        case kinds = "Kinds"
    }
}

private struct ReceiveReq: Encodable {
    let amountSat: Int64
    let memo: String
    enum CodingKeys: String, CodingKey {
        case amountSat = "AmountSat"
        case memo = "Memo"
    }
}

private struct DepositReq: Encodable {
    let amountSatHint: Int64
    enum CodingKeys: String, CodingKey {
        case amountSatHint = "AmountSatHint"
    }
}

private struct PrepareSendReq: Encodable {
    let invoice: String?
    let onchainAddress: String?
    let amountSat: Int64
    let note: String
    let maxFeeSat: Int64
    let sweepAll: Bool
    enum CodingKeys: String, CodingKey {
        case invoice = "Invoice"
        case onchainAddress = "OnchainAddress"
        case amountSat = "AmountSat"
        case note = "Note"
        case maxFeeSat = "MaxFeeSat"
        case sweepAll = "SweepAll"
    }
}

private struct SendPreparedReq: Encodable {
    let sendIntentID: String
    enum CodingKeys: String, CodingKey {
        case sendIntentID = "SendIntentID"
    }
}

private struct ListReq: Encodable {
    let view: String
    let pendingOnly: Bool
    let kinds: [String]
    let limit: Int64
    let offset: Int64
    enum CodingKeys: String, CodingKey {
        case view = "View"
        case pendingOnly = "PendingOnly"
        case kinds = "Kinds"
        case limit = "Limit"
        case offset = "Offset"
    }
}

private struct ExitReq: Encodable {
    let outpoint: String
    let destination: String
    let forceUnrollAck: String
    enum CodingKeys: String, CodingKey {
        case outpoint = "Outpoint"
        case destination = "Destination"
        case forceUnrollAck = "ForceUnrollAck"
    }
}

private struct ExitStatusReq: Encodable {
    let outpoint: String
    enum CodingKeys: String, CodingKey {
        case outpoint = "Outpoint"
    }
}

File: android/walletkit/src/main/java/engineering/lightning/wavewalletdk/client/WalletClient.kt
package engineering.lightning.wavewalletdk.client

import android.util.Base64
import engineering.lightning.wavewalletdk.mobile.Mobile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.channelFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** Thrown when an embedded-wallet call fails. Wraps the Go-side error message. */
class WalletException(message: String?) : Exception(message)

/**
 * An idiomatic Kotlin facade over the gomobile `Mobile` bindings.
 *
 * Every call is a `suspend` function that runs the blocking binding on
 * [Dispatchers.IO] and returns a typed model (or throws [WalletException]).
 * The wallet activity stream is exposed as a [Flow] that closes its underlying
 * subscription when collection stops. One embedded daemon runs per process, so
 * a single `WalletClient` is enough for the whole app.
 */
class WalletClient(private val json: Json = DEFAULT_JSON) {

  /** Whether the embedded daemon is currently running. */
  val isRunning: Boolean get() = Mobile.isRunning()

  /** Boot the embedded daemon and block until its gRPC channel is serving. */
  suspend fun start(config: WalletConfig) = io {
    Mobile.start(json.encodeToString(WalletConfig.serializer(), config))
  }

  /** Tear the daemon down. Idempotent. */
  suspend fun stop() = io { Mobile.stop() }

  /** Daemon readiness snapshot. */
  suspend fun getInfo(): Info = decode(io { Mobile.getInfo() })

  /** Wallet readiness, balance, and pending-activity count. */
  suspend fun status(): Status = decode(io { Mobile.status() })

  /** Wallet balance summary. */
  suspend fun balance(): Balance = decode(io { Mobile.balance() })

  /** Confirmed balance in satoshis, without decoding a JSON payload. */
  suspend fun confirmedBalanceSat(): Long = io { Mobile.confirmedBalanceSat() }

  /** Create or import the wallet. An empty mnemonic generates a fresh seed. */
  suspend fun createWallet(
    walletPassword: ByteArray,
    mnemonic: List<String> = emptyList(),
  ): CreateWalletResult {
    val req = CreateWalletReq(
      mnemonic = mnemonic,
      walletPassword = base64(walletPassword),
    )
    return decode(
      io { Mobile.createWallet(encode(CreateWalletReq.serializer(), req)) },
    )
  }

  /** Unlock an existing wallet. */
  suspend fun unlockWallet(walletPassword: ByteArray): UnlockWalletResult {
    val req = UnlockWalletReq(base64(walletPassword))
    return decode(
      io { Mobile.unlockWallet(encode(UnlockWalletReq.serializer(), req)) },
    )
  }

  /**
   * Open a Lightning invoice to receive [amountSat] into the wallet. Share the
   * returned invoice with the payer; the matching entry shows up on the
   * activity stream as it is paid.
   */
  suspend fun receiveLightning(amountSat: Long, memo: String = ""): ReceiveResult =
    decode(io { Mobile.receive(encode(ReceiveReq.serializer(), ReceiveReq(amountSat, memo))) })

  /**
   * Allocate a fresh on-chain boarding address to deposit into. The optional
   * hint lets the daemon size the boarding round; zero leaves it to the daemon.
   */
  suspend fun newDepositAddress(amountSatHint: Long = 0): DepositResult =
    decode(io { Mobile.deposit(encode(DepositReq.serializer(), DepositReq(amountSatHint))) })

  /**
   * Quote an outbound payment without moving funds. Set exactly one of [invoice]
   * or [onchainAddress]. The returned [PrepareSendResult] carries a single-use
   * intent id to pass to [send]. [sweepAll] drains every VTXO to the address.
   */
  suspend fun prepareSend(
    invoice: String? = null,
    onchainAddress: String? = null,
    amountSat: Long = 0,
    note: String = "",
    maxFeeSat: Long = 0,
    sweepAll: Boolean = false,
  ): PrepareSendResult = decode(
    io {
      Mobile.prepareSend(
        encode(
          PrepareSendReq.serializer(),
          PrepareSendReq(invoice, onchainAddress, amountSat, note, maxFeeSat, sweepAll),
        ),
      )
    },
  )

  /** Dispatch a prepared send, consuming the single-use intent id. */
  suspend fun send(sendIntentId: String): SendResult =
    decode(io { Mobile.sendPrepared(encode(SendPreparedReq.serializer(), SendPreparedReq(sendIntentId))) })

  /**
   * Pay a Lightning invoice in one call, skipping the explicit quote step. Use
   * [prepareSend] + [send] instead when you want to show the fee before paying.
   */
  suspend fun payInvoice(invoice: String, maxFeeSat: Long = 0): SendResult {
    val quote = prepareSend(invoice = invoice, maxFeeSat = maxFeeSat)
    return send(quote.sendIntentId)
  }

  /**
   * Stream wallet activity as a [Flow]. Each emission is one [Entry]; the flow
   * completes on a clean end-of-stream and fails with [WalletException] on a
   * real error. Cancelling collection closes the underlying subscription,
   * which unblocks the pull loop.
   */
  fun activity(
    includeExisting: Boolean = false,
    kinds: List<String> = emptyList(),
  ): Flow<Entry> = channelFlow {
    val req = encode(SubscribeReq.serializer(), SubscribeReq(includeExisting, kinds))
    val sub = Mobile.subscribe(req)

    val pump = launch(Dispatchers.IO) {
      try {
        while (isActive) {
          val next = try {
            sub.next()
          } catch (e: Exception) {
            if (isEndOfStream(e)) break else throw WalletException(e.message)
          }
          trySend(json.decodeFromString(Entry.serializer(), String(next)))
        }
        close()
      } catch (e: Throwable) {
        close(e)
      }
    }

    awaitClose {
      sub.close()
      pump.cancel()
    }
  }

  /**
   * List a unified wallet view: the merged activity history, the live VTXO
   * inventory, or the on-chain transaction history. Read the [ListResult] field
   * named by its `view`. `kinds`, `pendingOnly`, `limit`, and `offset` apply to
   * the activity view.
   */
  suspend fun list(
    view: ListView = ListView.ACTIVITY,
    kinds: List<String> = emptyList(),
    pendingOnly: Boolean = false,
    limit: Long = 0,
    offset: Long = 0,
  ): ListResult = decode(
    io {
      Mobile.list(
        encode(ListReq.serializer(), ListReq(view.wire, pendingOnly, kinds, limit, offset)),
      )
    },
  )

  /**
   * Exit a VTXO back to the chain. The daemon queues a cooperative leave by
   * default, paying out to [destination] (or a fresh backing-wallet address when
   * empty). To bypass cooperation and start a unilateral unroll, pass
   * [forceUnrollAck] = "I_KNOW_WHAT_I_AM_DOING"; it cannot be combined with
   * [destination]. Track progress with [exitStatus] or the activity stream.
   */
  suspend fun exit(
    outpoint: String,
    destination: String = "",
    forceUnrollAck: String = "",
  ): ExitResult = decode(
    io {
      Mobile.exit(encode(ExitReq.serializer(), ExitReq(outpoint, destination, forceUnrollAck)))
    },
  )

  /** Query the phase of an exit job for a VTXO outpoint. */
  suspend fun exitStatus(outpoint: String): ExitStatusResult =
    decode(io { Mobile.exitStatus(encode(ExitStatusReq.serializer(), ExitStatusReq(outpoint))) })

  /**
   * Stream only incoming payments (receives and deposits) as they arrive and
   * settle. A convenience over [activity] filtered to the credit kinds, useful
   * for a "you were paid" notification: watch for an [Entry] whose status
   * becomes "complete".
   */
  fun incomingPayments(includeExisting: Boolean = false): Flow<Entry> =
    activity(includeExisting = includeExisting, kinds = listOf("receive", "deposit"))

  // --- internals -----------------------------------------------------------

  private suspend inline fun <T> io(crossinline block: () -> T): T =
    try {
      withContext(Dispatchers.IO) { block() }
    } catch (e: WalletException) {
      throw e
    } catch (e: Exception) {
      throw WalletException(e.message)
    }

  private inline fun <reified T> decode(bytes: ByteArray): T =
    json.decodeFromString(String(bytes))

  private fun <T> encode(
    serializer: kotlinx.serialization.SerializationStrategy<T>,
    value: T,
  ): ByteArray = json.encodeToString(serializer, value).toByteArray()

  companion object {
    private val DEFAULT_JSON = Json {
      ignoreUnknownKeys = true
      explicitNulls = false
    }

    private fun base64(b: ByteArray): String = Base64.encodeToString(b, Base64.NO_WRAP)

    private fun isEndOfStream(e: Exception): Boolean =
      e.message?.contains("EOF", ignoreCase = true) == true
  }
}

@Serializable
private data class CreateWalletReq(
  @SerialName("Mnemonic") val mnemonic: List<String> = emptyList(),
  @SerialName("WalletPassword") val walletPassword: String,
)

@Serializable
private data class UnlockWalletReq(
  @SerialName("WalletPassword") val walletPassword: String,
)

@Serializable
private data class SubscribeReq(
  @SerialName("IncludeExisting") val includeExisting: Boolean = false,
  @SerialName("Kinds") val kinds: List<String> = emptyList(),
)

@Serializable
private data class ReceiveReq(
  @SerialName("AmountSat") val amountSat: Long,
  @SerialName("Memo") val memo: String = "",
)

@Serializable
private data class DepositReq(
  @SerialName("AmountSatHint") val amountSatHint: Long = 0,
)

@Serializable
private data class PrepareSendReq(
  @SerialName("Invoice") val invoice: String? = null,
  @SerialName("OnchainAddress") val onchainAddress: String? = null,
  @SerialName("AmountSat") val amountSat: Long = 0,
  @SerialName("Note") val note: String = "",
  @SerialName("MaxFeeSat") val maxFeeSat: Long = 0,
  @SerialName("SweepAll") val sweepAll: Boolean = false,
)

@Serializable
private data class SendPreparedReq(
  @SerialName("SendIntentID") val sendIntentId: String,
)

@Serializable
private data class ListReq(
  @SerialName("View") val view: String,
  @SerialName("PendingOnly") val pendingOnly: Boolean = false,
  @SerialName("Kinds") val kinds: List<String> = emptyList(),
  @SerialName("Limit") val limit: Long = 0,
  @SerialName("Offset") val offset: Long = 0,
)

@Serializable
private data class ExitReq(
  @SerialName("Outpoint") val outpoint: String,
  @SerialName("Destination") val destination: String = "",
  @SerialName("ForceUnrollAck") val forceUnrollAck: String = "",
)

@Serializable
private data class ExitStatusReq(
  @SerialName("Outpoint") val outpoint: String,
)
